libs-scala/resources/src/main/scala/com/digitalasset/resources/ExecutorServiceResourceOwner.scala

```diff
@@ -5,17 +5,51 @@ package com.digitalasset.resources
 
 import java.util.concurrent.{ExecutorService, TimeUnit}
 
-import scala.concurrent.{ExecutionContext, Future}
+import com.digitalasset.resources.ExecutorServiceResourceOwner._
+
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutorService, Future}
 
 class ExecutorServiceResourceOwner[T <: ExecutorService](acquireExecutorService: () => T)
     extends ResourceOwner[T] {
   override def acquire()(implicit executionContext: ExecutionContext): Resource[T] =
     Resource(
-      Future(acquireExecutorService()),
+      Future {
+        val executorService = acquireExecutorService()
+        executionContext match {
+          case context: ExecutionContextExecutorService =>
+            if (executorService == context) {
+              throw new CannotAcquireExecutionContext()
+            }
+            // Ugly, but important so that we make sure we're not going to end up in deadlock.
```

Comment:
So this ugly piece of code takes care when an execution context is wrapped via `ExecutionContext.fromExecutorService`, but it will not detect the problem if a different wrapper is used.

Alternatively, you could try to submit something to the given execution context and look at the executing thread's name to see whether they belong to the same execution context. Unfortunately, as there is no standard on how execution contexts name their threads, that's probably as brittle as the current mechanism.

Reply:
I think the comment is probably the wisest. We're just trying to make sure people don't do something mad; we can't prevent them all. :-)

If people write tests, they'll catch this themselves, of course.

Reply:
Added a lot more commentary.